pkg/util/ticker.go

```diff
@@ -0,0 +1,65 @@
+package util
+
+import (
+	"context"
+	"math/rand"
+	"time"
+)
+
+type Jitter struct {
+	base      time.Duration
+	deviation time.Duration
+}
+
+// NewJitter returns a Jitter object that creates durations with random jitter.
+func NewJitter(b time.Duration, d time.Duration) Jitter {
+	return Jitter{base: b, deviation: d}
+}
+
+// Duration returns a random duration from the base duration and +/- jitter
+func (j Jitter) Duration() time.Duration {
+	base := j.base - j.deviation
+	jitter := time.Duration(rand.Int63n(int64(float64(2 * j.deviation.Nanoseconds()))))
+	return base + jitter
+}
+
+type TickerWithJitter struct {
+	C    chan time.Time
+	ctx  context.Context
+	stop func()
+}
+
+func (t *TickerWithJitter) Stop() {
+	t.stop()
+}
+
+func (t *TickerWithJitter) start(d, dev time.Duration) {
+	j := NewJitter(d, dev)
+	timer := time.NewTimer(j.Duration())
+	defer timer.Stop()
+	for {
+		select {
+		case <-timer.C:
+			timer.Reset(j.Duration())
```

Comment:
hmmm does that mean you'll be calling "Duration()" to every iter, doesn't it? if so, i feel like it might get expensive. WDYT of calculating the randomness a single time when an ingester is being instantiated, but then to use the random factor on every iteration without reevaluating it?

Reply:
I explicitly do not want a fixed ticker interval.

Reply:
sorry I thought he meant to not rely only on an initial sleep but to rather apply the sleep/jitter on every evaluation